src/placeos-rest-api/controllers/openai/chat_manager.cr

```diff
@@ -0,0 +1,270 @@
+    end
+
+    private def save_history(chat_id : String, msg : OpenAI::ChatMessage)
+      save_history(chat_id, PlaceOS::Model::ChatMessage::Role.parse(msg.role.to_s), msg.content || "", msg.name, msg.function_call.try &.arguments)
+    end
+
+    private def build_prompt(chat : PlaceOS::Model::Chat, chat_payload : Payload?)
+      messages = [] of OpenAI::ChatMessage
+
+      if payload = chat_payload
+        messages << OpenAI::ChatMessage.new(role: :system, content: payload.prompt)
+        messages << OpenAI::ChatMessage.new(role: :system, content: "request function lists and call functions as required to fulfil requests.\n" +
+                                                                    "make sure to interpret and reply with the results of function calls.\n" +
+                                                                    "remember to only use valid capability ids, they can be found in this JSON:\n```json\n#{payload.capabilities.to_json}\n```" +
+                                                                    "id strings are case sensitive and must not be modified.")
+
+        messages.each { |m| save_history(chat.id.as(String), m) }
+      else
+        chat.messages.each do |hist|
+          func_call = nil
+          if hist.role.to_s == "function"
+            if name = hist.function_name
+              args = hist.function_args || JSON::Any.new(nil)
+              func_call = OpenAI::ChatFunctionCall.new(name, args)
+            end
+          end
+          messages << OpenAI::ChatMessage.new(role: OpenAI::ChatMessageRole.parse(hist.role.to_s), content: hist.content,
+            name: hist.function_name,
+            function_call: func_call
+          )
+        end
+      end
+
+      messages
+    end
+
+    private def driver_prompt(chat : PlaceOS::Model::Chat) : Payload?
+      resp, code = exec_driver_func(chat, LLM_DRIVER, LLM_DRIVER_CHAT, nil)
+      if code >= 200 && code <= 299
+        Payload.from_json(resp)
+      else
+        raise "error obtaining chat prompt: #{resp} (#{code})"
+      end
+    end
+
+    private def build_executor(chat, payload : Payload?)
+      executor = OpenAI::FunctionExecutor.new
+
+      description = if payload
+                      "You have the following capability list, described in the following JSON:\n```json\n#{payload.capabilities.to_json}\n```\n" +
+                        "if a request could benefit from these capabilities you can obtain the list of functions by providing the id string.\n" +
+                        "id strings are case sensitive and must not be modified."
+                    else
+                      "if a request could benefit from a capability you can obtain the list of functions by providing the id string\n" +
+                        "id strings are case sensitive and must not be modified."
+                    end
```

Comment:
The only thing I think that needs tweaking is this chat resume
It would be nice to have the same description with the capabilities embedded

So we need to make the `LLM.new_chat` request call for chat resume so the data can be available